pkg/sql/virtual_table.go

```diff
@@ -24,20 +25,95 @@ import (
 // nil). If there is an error, then return (nil, error).
 type virtualTableGenerator func() (tree.Datums, error)
 
+type virtualTableGeneratorResponse struct {
+	datums tree.Datums
+	err    error
+}
+
+var errRowsDone = errors.New("predefined error that the generator has been canceled")
+
+// setupGenerator takes in a worker that generates rows eagerly and transforms
+// it into a lazy row generator. It returns three functions:
+// * cleanup: Performs all cleanup. May only be called once.
+// * next: A handle that can be called to generate a row from the worker. Next
+//   cannot be called once cleanup has been called.
+func setupGenerator(
+	worker func(addRow func(...tree.Datum) error) error,
+) (cleanup func(), next virtualTableGenerator) {
+	allowWork, done, send := make(chan struct{}), make(chan struct{}), make(chan virtualTableGeneratorResponse)
+	cleanup = func() {
+		close(done)
+	}
+	addRow := func(datums ...tree.Datum) error {
+		select {
+		case <-done:
+			return errRowsDone
+		case send <- virtualTableGeneratorResponse{datums: datums}:
+		}
+
+		// Block until the next call to cleanup() or next(). This allows us to
+		// avoid issues with concurrent transaction usage if the worker is using
+		// a transaction. Otherwise, worker could proceed running operations after
+		// a call to next() has returned. That could result in the main operator
+		// chain using the transaction while the worker is also running. This
+		// makes it so that the worker can only run while next() is being called,
+		// which effectively gives ownership of the transaction usage over to the
+		// worker, and then back to the next() caller after it is done.
+		select {
+		case <-done:
+			return errRowsDone
+		case <-allowWork:
+		}
+		return nil
+	}
+
+	startGenerator := func() {
```

Comment:
could just inline say `go func() {` instead of the startGenerator declaration if you wanted to :P

Reply:
done